### PR TITLE
Minor authentication changes

### PR DESCRIPTION
**Pull request by Harsh14901**
## Features implemented
- Shifted from HS256 to RS256
- Changed verify-token endpoint to refresh-token endpoint
- Eliminated hardcoding of token names

**Note:** Presently the RSA key pair is being published on the repo, for ease of tesing. But later on during deployment, the keys will be definitely changed and removed from the repo and placed properly in config vars.

### CODE REVIEW

File: src/routes/auth.js [hunk 1]
```diff
@@ -1,13 +1,14 @@
...
 
 // post route to check validity of tokens, clients will hit this route.
-router.post('/verify-token', (req, res) => {
-    // extract token from post request headers
-    const token = req.header('access-token');
-    const rememberme = req.header('refresh-token');
+router.post('/refresh-token', (req, res) => {
+    // Extract tokens from request body
+    const token = req.body[tokenName];
+    const rememberme = req.body[rememberTokenName];
```
Comment (bicycleman15) [inline src/routes/auth.js:11]: we extract token from `req.header`, not `req.body`

Comment (bicycleman15) [reply src/routes/auth.js:11]: And I think naming convention, `access-token-name` and `refresh-token-name` is better.

Comment (bicycleman15) [reply src/routes/auth.js:11]: Also change every instance of `rememberme` variable to `refresh_token` maybe

File: src/routes/profile.js [hunk 2]
```diff
@@ -9,7 +10,8 @@ router.get('/settings', (req, res) => {
 
 router.post('/', (req, res) => {
     // extract token from cookie
-    const { token, rememberme } = req.cookies;
+    const token = req.cookies[tokenName];
+    const rememberme = req.cookies[rememberTokenName];
     if (!token) {
```
Comment (bicycleman15) [inline src/routes/profile.js:15]: same as discussed above

File: src/routes/profile.js [hunk 3]
```diff
@@ -18,16 +20,16 @@ router.post('/', (req, res) => {
...
 
     return verifyToken(token, res);
 });
 
 router.post('/logout', (req, res) => {
     try {
-        res.clearCookie('token');
-        res.clearCookie('rememberme');
+        res.clearCookie(tokenName);
+        res.clearCookie(rememberTokenName);
```
Comment (bicycleman15) [inline src/routes/profile.js:32]: same

File: src/utils/utils.js [hunk 4]
```diff
@@ -28,14 +30,14 @@ const createJWTCookie = (user, res, tokenName = 'token') => {
     });
 };
 
-const verifyToken = (token, res, tokenName = 'token') => {
+const verifyToken = (token, res, tokenName = keys.tokenName) => {
     try {
-        const decoded = verify(token, secretkey);
+        const decoded = verify(token, keys.publicKey);
```
Comment (bicycleman15) [inline src/utils/utils.js:35]: I think when you verify you need to also mention the algorithm you used to sign for added security, maybe look that up and add that as well

File: src/routes/auth.js (hunk 1 above)
Comment (Harsh14901) [reply src/routes/auth.js:11]: > we extract token from `req.header`, not `req.body`
@bicycleman15 
Afaik, the data in a POST request is sent in the body and not the headers. Will sending it in the headers have any advantage over sending it in the body?

Comment (bicycleman15) [reply src/routes/auth.js:11]: It's a convention to send token in header

Comment (bicycleman15) [reply src/routes/auth.js:11]: Have a look at this. https://stackoverflow.com/questions/53880867/is-there-any-security-difference-between-http-headers-and-http-body

Comment (Harsh14901) [reply src/routes/auth.js:11]: Yeah I understand, that is the case when you need to authenticate with an API endpoint, then we do send our authorization token in the `Authorization` header. The API endpoint then authorizes the **sender of the request** and then allows access to data that the endpoint is supposed to provide/operate upon. In our case the endpoint's task itself is to verify the token and the remember-me token. 

So in our example suppose we need to authenticate the SSO client (which we have not implemented) like yearbook, then we will provide an access token to yearbook which it will send in the POST request with `Authorization: Bearer <yearbook_token>`. and the tokens (`token` and `rememberme`) that need to be verified for the end user visiting yearbook will be sent in the body.

So basically what I am trying to say is that in our case the token that we are sending to the API endpoint **are NOT the tokens that authorize the sender of the request (yearbook)** to access the endpoint, but they themselves are the data that the endpoint is supposed to operate on.

I hope you get my point.

Comment (bicycleman15) [reply src/routes/auth.js:11]: Okay then, if you insist.
